yield_plugin/src/yield_plugin.cpp

```diff
@@ -403,106 +415,243 @@ namespace yield_plugin
-    RCLCPP_DEBUG_STREAM(nh_->get_logger(),vehicle_downtrack);
+    // save route Ids for faster access
+    for (auto llt: wm_->getRoute()->shortestPath())
+    {
+      route_llt_ids_.insert(llt.id());
+    }
 
-    for (auto i : rwol2.roadway_obstacles){
+    double earliest_collision_obj_time = std::numeric_limits<double>::max();
 
-      lanelet::BasicPoint2d point_o(i.object.pose.pose.position.x, i.object.pose.pose.position.y);
-      double object_down_track = wm_->routeTrackPos(point_o).downtrack;
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"),"External Object List (external_objects_) size: " << external_objects_.size());
 
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"object_down_track");
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),object_down_track);
-      
-      double dist_to_object = object_down_track - vehicle_downtrack;
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"object_down_track - vehicle_downtrack");
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),dist_to_object);
-      
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"i.object.velocity.twist.linear.x");
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),i.object.velocity.twist.linear.x);
-
-      if(current_velocity.linear.x > 0.0) {
-          RCLCPP_DEBUG_STREAM(nh_->get_logger(),"(object_down_downtrack - vehicle_downtrack)/current_velocity.linear.x");
-          RCLCPP_DEBUG_STREAM(nh_->get_logger(),dist_to_object/current_velocity.linear.x);
-
-          // Check to see if the object is in front of us and has a time-to-collision lower than our
-          // desired horizon
-          if(dist_to_object >= 0 &&
-            dist_to_object / current_velocity.linear.x < config_.collision_horizon) {
-              rwol_collision.push_back(i);
+    for (auto curr_obstacle : external_objects_)
+    {
+      RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Object's back time: " << std::to_string(rclcpp::Time(curr_obstacle.predictions.back().header.stamp).seconds()) << ", plan_start_time: " << std::to_string(plan_start_time.seconds()));
+
+      // do not process outdated objects
+      if (rclcpp::Time(curr_obstacle.predictions.back().header.stamp) > plan_start_time)
+      {
+        carma_perception_msgs::msg::PredictedState curr_state;
+        // artificially include current position as one of the predicted states
+        curr_state.header.stamp = curr_obstacle.header.stamp;
+        curr_state.predicted_position.position.x = curr_obstacle.pose.pose.position.x;
+        curr_state.predicted_position.position.y = curr_obstacle.pose.pose.position.y;
+        curr_state.predicted_velocity.linear.x = curr_obstacle.velocity.twist.linear.x;
+        new_list.push_back(curr_state);
+        new_list.insert(new_list.end(), curr_obstacle.predictions.begin(), curr_obstacle.predictions.end());
+
+        auto collision_time = detect_collision_time(original_tp, new_list, config_.intervehicle_collision_distance);
+        
+        if (collision_time)
```

Comment:
Using `collision_time != boost::none` or `collision_time != std::nullopt` is more readable.